### PR TITLE
feat: password modals for private crowdactions

### PR DESCRIPTION
**Pull request by yasht01**

### CODE REVIEW

File: lib/domain/core/i_settings_repository.dart [hunk 1]
```diff
@@ -5,4 +5,10 @@ abstract class ISettingsRepository {
 
   // ignore: avoid_positional_boolean_parameters
   Future<void> setWasUserOnboarded({required bool wasOnboarded});
+
+  //* Get list of private crowdaction IDs the user has access to.
```
Comment (Xazin) [inline lib/domain/core/i_settings_repository.dart:9]: Shouldn't this be in user repository?

File: lib/domain/core/i_settings_repository.dart [hunk 2]
```diff
@@ -5,4 +5,10 @@ abstract class ISettingsRepository {
...
+  Future<List<String>> getCrowdActionAccessList();
+
+  //* Add _crowdactionId_ to list of strings in shared preferences.
```
Comment (Xazin) [inline lib/domain/core/i_settings_repository.dart:12]: Also user repository maybe?

File: lib/domain/core/i_settings_repository.dart (hunk 1 above)
Comment (yasht01) [reply lib/domain/core/i_settings_repository.dart:9]: @Xazin are you referring to all the four functions or only to these two:
- `Future<void> setWasUserOnboarded({required bool wasOnboarded});`
- `Future<List<String>> getCrowdActionAccessList();`

Comment (yasht01) [reply lib/domain/core/i_settings_repository.dart:9]: Also, I saw that most of the functions in the `UserRepository` are related to authentication so was confused as to what it was exactly meant for.

File: lib/presentation/crowd_action/pages/widgets/in_spotlight_header.dart [hunk 3]
```diff
@@ -97,15 +99,24 @@ class _InSpotLightHeaderState extends State<InSpotLightHeader> {
                                       ),
                                     ),
                                   ),
-                                  ...crowdAction.toChips()
+                                  ...crowdAction.toChips(),
+                                  if (crowdAction.passwordJoin != "")
```
Comment (Xazin) [inline lib/presentation/crowd_action/pages/widgets/in_spotlight_header.dart:103]:
```suggestion
                                  if (crowdAction.passwordJoin.isNotEmpty)
```

File: lib/presentation/crowd_action/pages/widgets/in_spotlight_header.dart [hunk 4]
```diff
@@ -97,15 +99,24 @@ class _InSpotLightHeaderState extends State<InSpotLightHeader> {
...
                                 ],
                                 participants: crowdAction.topParticipants,
                                 totalParticipants: crowdAction.participantCount,
-                                onTap: () => context.router.push(
-                                  CrowdActionDetailsRoute(
-                                    crowdAction: crowdAction,
-                                  ),
-                                ),
+                                onTap: () {
+                                  if (crowdAction.passwordJoin != "") {
```
Comment (Xazin) [inline lib/presentation/crowd_action/pages/widgets/in_spotlight_header.dart:113]:
```suggestion
                                  if (crowdAction.passwordJoin.isNotEmpty) {
```

File: lib/presentation/home/widgets/current_upcoming_layout.dart [hunk 5]
```diff
@@ -84,17 +86,18 @@ class _CurrentAndUpcomingLayoutState extends State<CurrentAndUpcomingLayout> {
...
 
-  // Widget _noCrowdAction() {
-  //   return const Text('No crowdactions to show');
-  // }
-
   Widget _fetched(List<CrowdAction> fetchedData) {
     return Column(
       children: fetchedData
           .map(
             (e) => GestureDetector(
```
Comment (Xazin) [inline lib/presentation/home/widgets/current_upcoming_layout.dart:93]: `e` needs a better name

File: lib/presentation/home/widgets/current_upcoming_layout.dart [hunk 6]
```diff
@@ -84,17 +86,18 @@ class _CurrentAndUpcomingLayoutState extends State<CurrentAndUpcomingLayout> {
...
-              onTap: () =>
-                  context.router.push(CrowdActionDetailsRoute(crowdAction: e)),
+              onTap: () {
+                if (e.passwordJoin != "") {
```
Comment (Xazin) [inline lib/presentation/home/widgets/current_upcoming_layout.dart:95]:
```suggestion
                if (e.passwordJoin.isNotEmpty) {
```

File: lib/presentation/home/widgets/current_upcoming_layout.dart [hunk 7]
```diff
@@ -122,12 +125,20 @@ class _CurrentAndUpcomingLayoutState extends State<CurrentAndUpcomingLayout> {
...
+                                    MainAxisAlignment.spaceBetween,
+                                children: [
+                                  const AccentChip(
+                                    text: "Sign up now",
+                                    leading: Icon(
+                                      Icons.check,
+                                      color: Colors.white,
+                                    ),
+                                  ),
+                                  if (e.passwordJoin != "")
```
Comment (Xazin) [inline lib/presentation/home/widgets/current_upcoming_layout.dart:139]:
```suggestion
                                  if (e.passwordJoin.isNotEmpty)
```

File: lib/presentation/home/widgets/password_modal.dart [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+            backgroundColor: _disableButton ? kPrimaryColor200 : kAccentColor,
+            minRadius: 30,
+            child: IconButton(
+              onPressed: () {
+                if (_disableButton == true) {
+                  return;
+                } else {
+                  _validatePassword();
+                }
+              },
```
Comment (Xazin) [inline lib/presentation/home/widgets/password_modal.dart:129]:
```suggestion
              onPressed: !_disableButton ?  () {
                  _validatePassword();
                } : null,
```

File: lib/presentation/home/widgets/password_modal.dart [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+      setState(() {
+        _validated = true;
+      });
+      addCrowdActionAccess();
+      context.router.pop();
+      context.router.push(
+        CrowdActionDetailsRoute(
+          crowdAction: widget.crowdAction,
+        ),
+      );
```
Comment (Xazin) [inline lib/presentation/home/widgets/password_modal.dart:156]: Can't we use `context.router.replace()` ?

File: lib/presentation/home/widgets/password_modal.dart [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        _validated = false;
+      });
+    }
+  }
+
+  Future<void> addCrowdActionAccess() async {
+    final _settingsRepository = getIt<ISettingsRepository>();
+    await _settingsRepository.addCrowdActionAccess(
+        crowdActionId: widget.crowdAction.crowdactionID);
+  }
```
Comment (Xazin) [inline lib/presentation/home/widgets/password_modal.dart:168]: All of this logic, together with all `setState` that we have, should be encapuslated in a BLOC, so we don't have application logic in presentation layer.

**Review by Xazin — APPROVED** [review]

**Pull request merged**